clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -306,6 +318,42 @@ static int64_t getIntExprValue(const Expr *E, ASTContext &Ctx) {
   return E->getIntegerConstantExpr(Ctx)->getSExtValue();
 }
 
+// Collect function attributes related to SYCL
+void CollectSYCLAttributes(Sema &S, FunctionDecl *FD,
+                           llvm::SmallPtrSet<Attr *, 4> &Attrs,
+                           bool directly_called = true) {
+  if (auto *A = FD->getAttr<IntelReqdSubGroupSizeAttr>())
```

Comment:
I think this can be simplified into:
```
llvm::copy_if(FD->getAtts(), std::inserter(Attrs), [](Attr *A) {
  return isa<IntelReqdSubGroupSizeAttr, ReqdWorkGroupSizeAttr, SYCLIntelKernelArgsRestrictAttr, etc>(A);
});
```
This is functionally different in that the approach in the patch will only add the attribute to the container *once* whereas the looping approach will get all of the attributes if any of them have been duplicated.

It may be worth it to consider splitting this into two functions. One that collects all the copies of everything but `SYCLIntelUseStallEnableClustersAttr`, and one that uses `FD->specific_attrs<SYCLIntelUseStallEnableClustersAttr>()` to loop over just those attributes and handle them.

Reply:
Thinking on this a bit more... it would be best if we were able to use the table generator to implement the logical guts of this function. This way, we could add something to each attribute definition in `Attr.td`, like `let CopyToKernel = 1;`, to signify which attributes opt into this behavior. Then `ClangAttrEmitter.cpp` could be modified to spit out a new .inc file that we `#include` within the `llvm::copy_if()` lambda. On the assumption that most attributes will not need custom logic like `SYCLIntelUseStallEnableClustersAttr` does, this could make it easier to opt attributes into this functionality in the future. WDYT? (To be clear, I'm not suggesting this tablegen work as part of this patch but instead as a follow-up refactoring. So I'm mostly wondering if that idea is worth opening an issue for or not.)

Reply:
I am unable to get the llvm::copy_if call to compile. Is there an example of similar usage elsewhere?

Reply:
Ugh, I didn't realize that `llvm::SmallPtrSet<>` doesn't have the correct interface for its `insert()` method -- it doesn't have one accepting an iterator hint. I realized that after noticing that my call to `std::inserter()` left off the iterator argument (sorry about that). See https://github.com/intel/llvm/blob/sycl/llvm/include/llvm/ADT/SmallPtrSet.h#L364

I actually think this is a deficiency in the upstream implementation. I think `SmallPtrSetImpl` should be given a function like:
```
std::pair<iterator, bool> insert(iterator, PtrType Ptr) {
  return insert(Ptr);
}
```
where the hint is accepted and ignored.

However, I can understand not wanting to do that as part of this patch, so an alternative suggestion is to do the loop copy manually:
```
for (Attr *A : FD->getAttrs()) {
  if (isa<IntelReqdSubGroupSizeAttr, ReqdWorkGroupSizeAttr, SYCLIntelKernelArgsRestrictAttr, etc>(A))
    Attrs.insert(A);
}
```

Reply:
OK, thanks for the tip.

Reply:
FWIW: https://reviews.llvm.org/D96134

Reply:
Thanks for following up on the need for an inserter.